tgg/annotations/get_ensembl_db_info.py

```diff
@@ -22,53 +24,168 @@
 homo_sapiens_variation_103_38
 """
 
-def get_canonical_transcripts(database=CURRENT_ENSEMBL_DATABASE):
-    gene_id_to_canonical_transcript_id = {}
-    with pymysql.connect(host="useastdb.ensembl.org", user="anonymous", database=database) as conn:
-        with conn.cursor() as c:
-            columns = ["gene.stable_id", "transcript.stable_id"]
-            columns_str = ", ".join(columns)
-            c.execute(f"SELECT {columns_str} FROM gene LEFT JOIN transcript ON canonical_transcript_id = transcript_id")
 
-            for row in c:
-                d = dict(zip(columns, row))
-                gene_id = d['gene.stable_id']
-                transcript_id = d['transcript.stable_id']
-                if gene_id in gene_id_to_canonical_transcript_id:
-                    other_id = gene_id_to_canonical_transcript_id[gene_id]
-                    raise Exception(f"{gene_id} has more than 1 canonical transcript: {transcript_id}, {other_id}")
-                gene_id_to_canonical_transcript_id[gene_id] = transcript_id
+def get_gene_id_to_transcript_metadata(
+        database=CURRENT_ENSEMBL_DATABASE,
+        only_protein_coding=False,
+        only_canonical_transcripts=False):
+    """Retrieves a dictionary containing gene_id => a list of dictionaries each of which
+    contains information about one transcript that belongs to that gene.
 
-    return gene_id_to_canonical_transcript_id
+    Args:
+        database (str): The Ensembl database name (eg. "homo_sapiens_core_107_38")
+        only_protein_coding (bool): If True, only return protein-coding genes and protein-coding transcripts
+        only_canonical_transcripts (bool): If True, only return canonical transcripts
 
+    Return:
+        dict: mapping ENSG id string to a list of dictionaries where each dictionary contains metadata fields
+    """
 
-def get_gene_created_modified_dates(database=CURRENT_ENSEMBL_DATABASE):
-    gene_id_to_dates = {}
+    gene_id_to_transcript_id = collections.defaultdict(list)
     with pymysql.connect(host="useastdb.ensembl.org", user="anonymous", database=database) as conn:
-        with conn.cursor() as c:
-            columns = ["stable_id", "created_date", "modified_date"]
+        with conn.cursor() as cursor:
+            if only_canonical_transcripts:
+                join_clause = "canonical_transcript_id = transcript_id"
+            else:
+                join_clause = "transcript.gene_id = gene.gene_id"
+
+            columns = [
+                "gene.stable_id",
+                "gene.biotype",
+                "gene.created_date",
+                "gene.modified_date",
+
```

Comment:
rather than a newline here, would it be helpful to make this something like
```
columns = [
                # Gene fields
                "gene.stable_id",
                "gene.biotype",
                "gene.created_date",
                "gene.modified_date",
                # Transcript fields
                "transcript.stable_id",
                "transcript.biotype",
                "transcript.created_date",
                "transcript.modified_date",
```